docs/development/zod.md

```diff
@@ -0,0 +1,291 @@
+
+External data being queried by Renovate can be very complex, while we may only need a subset of fields for our use cases.
+Avoid over-specifying schemas, and instead extract only the minimum necessary fields.
+If we don't include them, the surface of the contract between external data source is minimal, meaning less errors to fix in the future.
+
+Example of **incorrect** usage if we only care about width, height and length of a box:
+
+```ts
+const Box = z.object({
+  width: z.number(),
+  height: z.number(),
+  length: z.number(),
+  color: z.object({
+    red: z.number(),
+    green: z.number(),
+    blue: z.number(),
+  })
+  weight: z.number(),
+});
+
+const { width, height, length } = Box.parse(input);
+const volume = width * height * length;
+```
+
+Example of **correct** usage:
+
+```ts
+const Box = z.object({
+  width: z.number(),
+  height: z.number(),
+  length: z.number(),
+});
+
+const { width, height, length } = Box.parse(input);
+const volume = width * height * length;
+```
+
+### Use `Json`, `Yaml` and `Toml` for string parsing
+
+Sometimes we need to perform additional step such as `JSON.parse()` before validation of the data structure.
+Use helpers from `schema-utils.ts` for this purpose.
+
+Here is an **incorrect** way to parse from string:
+
+```ts
+try {
+  const json = JSON.parse(input);
+  return ApiResult.parse(json);
+} catch (e) {
+  return null;
+}
+```
+
+The **correct** way to parse from string:
+
+```ts
+Json.pipe(ApiResult).parse(input);
+```
+
+### Use `.transform()` method to process parsed data
+
+Schema validation helps to be more confident with types during downstream data transformation.
+
+You can go even further and perform some transformations as the part of schema itself.
+
+This is an example of **undesired** data transformation:
+
+```ts
+const Box = z.object({
+  width: z.number(),
+  height: z.number(),
+  length: z.number(),
+});
+
+const { width, height, length } = Box.parse(input);
+const volume = width * height * length;
+```
+
+Instead, use more idiomatic `.tranform()` method:
+
+```ts
+const Volume = z
+  .object({
+    width: z.number(),
+    height: z.number(),
+    length: z.number(),
+  })
+  .transform(({ width, height, length }) => width * height * length);
+
+Volume.parse({
+  width: 10,
+  height: 20,
+  length: 125,
+});
+```
+
+### Separate validation from transformation
+
+When parsing third party data, we are typically doing the following:
+
+- Validating that the data is correct/sufficient to use
+- Transforming it into a standardized format for internal usage
+
+Although it's not a strict requirement, your code will be cleaner if you perform the validation step first and then follow with transformation.
+
+### Stick to permissive behavior when possible
+
+Zod schemas are strict, and even some insufficient field is incorrect, the whole data will be treated like malformed.
+This could lead to cases when Renovate could've continued processing, but didn't.
```

Comment:
```suggestion
Zod schemas are strict, which means that if some field is wrong, or missing data, then the whole dataset is considered malformed.
Because Renovate uses Zod, it would then abort processing, even if we want Renovate to continue processing!
```